---
title: "2019_0115_fresh_start"
author: "Kirk Amundson"
date: "2019_0115"
output: html_notebook
---

TODO:

1. Define criteria for filtering out monomorphic bins. Either a percentage or average threshold will work.
   Note, in a typical mapping context, alleles this rare would be considered severely distored and removed on the
   basis of a chi square test or similar. Done
2. Loops for nonredundantly comparing alleles at all polymorphic loci. Done
3. Generate input data structure with lapply call. Doneish.
4. Multiple test correction. 

```{r}
library(tidyverse)
```

```{r}
# add monomorphic filter as high up as possible to make compute faster
dosage_genos <- read_tsv("2019_0107_LOP_250k_dosage_genotypes_40percband.tsv", col_names = T) %>% 
  mutate(bin = paste(chrom, start, end, sep = "_"))
  # filter(bin %in% c("chr01_0_250000", "chr01_250000_500000")) # training wheels, gradually ease off of them
head(dosage_genos)
```

```{r}
# progressively ease up on the training wheels
# dosage_genos_training <- dosage_genos %>% 
#   filter(bin %in% unique(dosage_genos$bin)[1:10])
```

```{r}
# filter out affected chromosome for each trisomic. These have been dealt with already.
dosage_genos_trifilt <-  dosage_genos %>% 
  filter(!(sample == "2x_LOP868_238" & chrom == "chr01")) %>%
  filter(!(sample == "2x_LOP868_259" & chrom == "chr02")) %>%
  filter(!(sample == "2x_LOP868_292" & chrom == "chr02")) %>%
  filter(!(sample == "2x_LOP868_452" & chrom == "chr02")) %>%
  filter(!(sample == "2x_LOP868_460" & chrom == "chr02")) %>%
  filter(!(sample == "2x_LOP868_1157" & chrom == "chr02")) %>%
  filter(!(sample == "2x_LOP868_1190" & chrom == "chr02")) %>%
  filter(!(sample == "2x_LOP868_373" & chrom == "chr03")) %>%
  filter(!(sample == "2x_LOP868_028" & chrom == "chr04")) %>%
  filter(!(sample == "2x_LOP868_433" & chrom == "chr04")) %>%
  filter(!(sample == "2x_LOP868_293" & chrom == "chr05")) %>%
  filter(!(sample == "2x_LOP868_363" & chrom == "chr05")) %>%
  filter(!(sample == "2x_LOP868_172" & chrom == "chr06")) %>%
  filter(!(sample == "2x_LOP868_065" & chrom == "chr07")) %>%
  filter(!(sample == "2x_LOP868_272" & chrom == "chr07")) %>%
  filter(!(sample == "2x_LOP868_108" & chrom == "chr08")) %>%
  filter(!(sample == "2x_LOP868_488" & chrom == "chr08")) %>%
  filter(!(sample == "2x_LOP868_397" & chrom == "chr10")) %>%
  filter(!(sample == "2x_LOP868_373" & chrom == "chr02")) %>%
  filter(!(sample == "2x_LOP868_128" & chrom == "chr12"))
```

```{r}
# split by bin, did this to get contingency tables of dosage genotypes
dosage_genos_split <- split(dosage_genos_trifilt, f=dosage_genos_trifilt$bin)
```

```{r}
# remove outlier individuals from data frame to be parsed while constructing fisher contingency tables
# this seems needlessly complicated...

# returns a list of named vectors. names of list are bin names. names of vectors are outlier dosage allele in that bin
outliers <- lapply(dosage_genos_split, function(x) table(x$dosage.gt)[which(table(x$dosage.gt) <= 3)])

# traverse outliers, parsing out bins and outlier dosage allele in each
z <- list()

# loop populates list of outlier bins and dosage alleles
for (i in 1:length(outliers)) {
  if (length(outliers[[i]]) >= 1) { # needs to skip bins without rows
  z[[i]] <- data.frame('dosage.gt' = names(outliers[[i]]),
             'bin' = names(outliers)[[i]])
  }
}

# bind rows to get single data frame of bin-allele combinations of all outliers
z.done <- bind_rows(z) %>% 
  mutate(allele = paste(bin, dosage.gt, sep = "_"))

# using z.done, remove outlier bin-allele combinations from table of dosage alleles by individual.
dosage_genos_outlierfilt <- dosage_genos_trifilt %>% 
  mutate(allele = paste(bin, dosage.gt, sep = "_")) %>% 
  filter(!allele %in% z.done$allele)

# did we filter anything?
nrow(dosage_genos_trifilt)
nrow(dosage_genos_outlierfilt) # yes

# parsable data structure for nested loops is stored in dosage_genos_outlierfilt
```

```{r}
# filter out other unique dosage variants. May need to adjust thresholding due to what I'm guessing are undetected tetraploids in the population
outliers_exclude <- lapply(dosage_genos_split, function(x) table(x$dosage.gt)[which(table(x$dosage.gt) > 3)]) # drops outliers, monomorphic still here
# outliers_exclude
```

```{r}
# what's left? This provides the scope of the analysis: the bins that are polymorphic
# lifted from other scripts, may need to develop this further
test <- unlist(lapply(outliers_exclude, function(x) length(x)))
# head(test)
# length(unique(names(which(test >= 2))))
dosage_variable_bins <- unique(names(which(test >= 2)))
head(dosage_variable_bins)
dosage_invariable_bins <- unique(names(which(test < 2)))
head(dosage_invariable_bins)
```

```{r}
monomorph_exclude <- outliers_exclude[which(names(outliers_exclude) %in% dosage_variable_bins)]
head(monomorph_exclude)
```

```{r}
# start with an example case
# geno_summaries <- lapply(dosage_genos_split, function(x) table(x$dosage.gt)) # outliers_exclude takes the place of this
# geno_summaries[[1]]
# names(geno_summaries[[1]])
# names(geno_summaries)[1]

# first example data frame
# example <- data.frame("bin" = names(geno_summaries)[1],
#                       geno_summaries[[1]]) %>% 
#   mutate(Var1 = as.character(Var1)) %>% 
#   mutate(bin_allele = paste(bin, Var1, sep = "_"))
# example

# second example data frame
# example2 <- data.frame("bin" = names(geno_summaries)[2],
#                        geno_summaries[[2]]) %>% 
#   mutate(Var1 = as.character(Var1)) %>% 
#   mutate(bin_allele = paste(bin, Var1, sep = "_"))
# example2

# build Fisher 2x2 contigency tables by comparison of both example data frames
# for (i in 1:nrow(example)) { # becomes xth entry in geno list
#   for (j in 1:nrow(example2)) { # becomes x+1th entry in geno list
#     observed.in <- example2[j,3] # hereafter x+1th entry in geno list, whatever variable I end up using for that.
#     observed.out <- sum(example2$Freq) - observed.in
#     expected <- example2[j,3] / 2
#     fisher.table <- matrix(c(observed.in, observed.out, expected, expected), nrow = 2)
#     list.name <- paste(example$bin_allele[i], example2$bin_allele[j], sep = " ")
#     fisher.list[[list.name]] <- fisher.table
#   }
# }
```

```{r}
# # next, build a more generalizable case using lists
# geno.list <- list() # stores genotype count data for each nonoverlapping bin
# 
# # build two examples as before, but this time have them be in the same list
# geno.list[[1]] <- data.frame("bin" = names(geno_summaries)[1], geno_summaries[[1]]) %>%
#   mutate(Var1 = as.character(Var1)) %>%
#   mutate(bin_allele = paste(bin, Var1, sep = "_"))
# geno.list[[1]]
# 
# # a second listy example
# geno.list[[2]] <- data.frame("bin" = names(geno_summaries)[2], geno_summaries[[2]]) %>%
#   mutate(Var1 = as.character(Var1)) %>%
#   mutate(bin_allele = paste(bin, Var1, sep = "_"))
# geno.list[[2]]
```

```{r}
# Generalizable conversion of genotype summary data to list of summary data
geno.list <- list() # stores genotype count data for each nonoverlapping bin

# work in progress chunk for generating geno.list using lapply call or loop
# for (i in 1:length(geno_summaries)) { 
for (i in 1:length(monomorph_exclude)) { 
  geno.list[[i]] <- data.frame("bin" = names(monomorph_exclude)[i], monomorph_exclude[[i]]) %>% 
    mutate(Var1 = as.character(Var1)) %>% 
    mutate(bin_allele = paste(bin, Var1, sep = "_"))
}
```

```{r}
# # sausage making to overhaul fisher exact comparison to be more like Luca's
# 
# # need to find subset of individuals that have the matching genotype in the target bin
# # then, the question is whether this subset is randomly distributed amongst the alleles in the target bin or not?
# 
# # keep in mind i would've locked in the allele comparison by this point
# 
# source.bin <- "chr01_1000000_1250000"
# source.cluster <- 1
# 
# target.bin <- "chr01_0_250000"
# target.cluster <- 1
# 
# # define individuals in source bin
# match.source.allele <- dosage_genos_outlierfilt %>%
#   filter(bin == source.bin & dosage.gt == source.cluster)
# nrow(match.source.allele)
# 
# # look at same individuals in target bin
# target.allele.dist <- dosage_genos_outlierfilt %>%
#   filter(bin == target.bin & sample %in% match.source.allele$sample)
# 
# g <- table(target.allele.dist$dosage.gt)
# 
# target.cluster.in.count <- g[which(names(g) == target.cluster)]
# target.cluster.out.count <- sum(g) - target.cluster.in.count
# target.cluster.null <- round(mean(g))
# 
# target.cluster.in.count
# target.cluster.out.count
# target.cluster.null
# 
# matrix(c(target.cluster.in.count, target.cluster.out.count, target.cluster.null, target.cluster.null), nrow = 2)
# 
# # not sure how this is going to handle multialleleic cases. There are a lot of them...
```

```{r}
# build fisher comparisons for all pairwise allele comparisons between two polymorphic loci
fisher.list <- list() # stores 2x2 contingency tables, lapply call fisher exact test at the end.

# nonredundant all by all pairwise comparison of data frames in list. each fisher contingency table saved as a named entry in the list.
for (x in seq(1, length(geno.list)-1)) { # loop through all except last bins. these are "source"
  # print(x)
  for (y in seq(x+1, length(geno.list))) { # all pairwise "target" bins, given the same source bin
    # print(y)
    for (i in 1:nrow(geno.list[[x]])) { # loop all alleles in source bin at hand
       for (j in 1:nrow(geno.list[[y]])) { # loop all alleles in target bin at hand
         source.bin <- geno.list[[x]][i,1]
         source.gt  <- geno.list[[x]][i,2]
         target.bin <- geno.list[[y]][j,1]
         target.gt  <- geno.list[[y]][j,2]
         # define individuals in source bin
         match.source.allele <- dosage_genos_outlierfilt %>%
            filter(bin == source.bin & dosage.gt == source.gt)
         target.allele.dist <- dosage_genos_outlierfilt %>%
            filter(bin == target.bin & sample %in% match.source.allele$sample) # why is this failing? It worked before!
         g <- table(target.allele.dist$dosage.gt)
         target.cluster.in.count <- g[which(names(g) == target.gt)]
         target.cluster.out.count <- sum(g) - target.cluster.in.count
         target.cluster.null <- round(mean(g))
         fisher.table <- matrix(c(target.cluster.in.count, target.cluster.out.count, target.cluster.null, target.cluster.null), nrow = 2)
         list.name <- paste(geno.list[[x]]$bin_allele[i], geno.list[[y]]$bin_allele[j], sep = " ")
         fisher.list[[list.name]] <- fisher.table
      }
    }
  }
}
# fisher.list 
```

```{r}
# TODO: Parse Fisher exact output, return as flat table so I can finish making new correlation matrix
fisher.price <- lapply(fisher.list, function(x) fisher.test(x))
fisher.price[[1]]
```

```{r}
# init empty data frame
fisher.df <- data.frame("src" = rep(NA, length(fisher.list)),
                        "trgt" = rep(NA, length(fisher.list)),
                        "pval" = rep(NA, length(fisher.list)))

# populate
for (i in 1:length(fisher.price)) {
  fisher.df$src[i]  <- str_split(names(fisher.price)[i], " ")[[1]][1]
  fisher.df$trgt[i] <- str_split(names(fisher.price)[i], " ")[[1]][2]
  fisher.df$pval[i] <- fisher.price[[i]]$p.value
}

# a few more cleanups
fisher.df.tidy <- fisher.df %>% 
  separate(src, into = paste0("src.", c("chrom", "start", "end", "gt")), remove = F, sep= "_") %>% 
  separate(trgt, into = paste0("trgt.", c("chrom", "start", "end", "gt")), remove = F, sep = "_") %>% 
  mutate(src.bin = str_replace_all(src, "(.+)_[0-9]$", "\\1")) %>% 
  mutate(trgt.bin = str_replace_all(trgt, "(.+)_[0-9]$", "\\1")) 
  
  # arrange(src.chrom, src.start, src.end, trgt.chsrc.gt, trgt.gt)
head(fisher.df.tidy)
```

```{r}
fisher.df.tidy$src.bin.num <- rep(NA, nrow(fisher.df.tidy))
fisher.df.tidy$trgt.bin.num <- rep(NA, nrow(fisher.df.tidy)) 

for (i in 1:nrow(fisher.df.tidy)) {
  fisher.df.tidy$src.bin.num[i] <- which(unique(dosage_genos$bin) == fisher.df.tidy$src.bin[i])
  fisher.df.tidy$trgt.bin.num[i] <- which(unique(dosage_genos$bin) == fisher.df.tidy$trgt.bin[i])
}
```

```{r}
View(fisher.df.tidy)
```

```{r}
# make glorious figure
ggplot(fisher.df.tidy, aes(x = src.bin.num, y = trgt.bin.num, color = pval)) +
  geom_tile()
```

```{r}
# nonredundant all by all pairwise comparison of data frames in list. each fisher contingency table saved as a named entry in the list.
# for (x in seq(1, length(geno.list)-1)) { # loop through all except last bins. these are "source"
#   # print(x)
#   for (y in seq(x+1, length(geno.list))) { # all pairwise "target" bins, given the same source bin
#     # print(y)
#     for (i in 1:nrow(geno.list[[x]])) { # loop all alleles in source bin at hand
#        for (j in 1:nrow(geno.list[[y]])) { # loop all alleles in target bin at hand
#          observed.in <- geno.list[[y]][i,3] # this has to change to match the sample lookup written above. Before I change too many things, leave this code block for posterity
#          # print(observed.in)
#          observed.out <- sum(geno.list[[y]]$Freq) - observed.in
#          # print(observed.out)
#          expected <- geno.list[[y]][j,3]
#          # print(expected)
#          fisher.table <- matrix(c(observed.in, observed.out, expected, expected), nrow = 2)
#          list.name <- paste(geno.list[[x]]$bin_allele[i], geno.list[[y]]$bin_allele[j], sep = " ")
#          fisher.list[[list.name]] <- fisher.table
#          # print(list.name)
#       }
#     }
#   }
# }
# fisher.list
```

```{r}
# once fisher list has been constructed, do fisher exact test on all tables in fisher.list with an lapply call
# lapply(fisher.list, fisher.test)
```

```{r}
# for visualization, plot 2-D points based on source and target bins. Color by inter-vs intrachromosomal, as well as by
# significance threshold.
```
