---
title: "2019_0115_fresh_start"
author: "Kirk Amundson"
date: "2019_0115"
output: html_notebook
---

TODO:

1. Define criteria for filtering out monomorphic bins. Either a percentage or average threshold will work.
   Note, in a typical mapping context, alleles this rare would be considered severely distored and removed on thes
   basis of a chi square test or similar. Done
2. Loops for nonredundantly comparing alleles at all polymorphic loci. Done
3. Generate input data structure with lapply call. Doneish.
4. Multiple test correction. I'm going to do Benjamini-Hochberg FDR correction after getting all p-values,
   so this is getting tabled for now.
5. Try to speed up initialization of count data. This set of four nested for loops is taking way too long to run.

```{r}
library(tidyverse)
```

```{r}
dosage_genos <- read_tsv("2019_0107_LOP_250k_dosage_genotypes_40percband.tsv", col_names = T) %>% 
  mutate(bin = paste(chrom, start, end, sep = "_")) %>% 
  filter(chrom %in% c("chr08", "chr09")) # putting the training wheels back on
  # filter(bin %in% c("chr01_0_250000", "chr01_250000_500000")) # training wheels
head(dosage_genos)
```

2. Optional: Filter out only the excess chromosome in each trisomic individual.
Note, I don't use this data structure in the rest of hte analysis. Left for posterity.

```{r}
# filter out affected chromosome for each trisomic. These have been dealt with already.
# dosage_genos_trifilt <-  dosage_genos %>% 
#   filter(!(sample == "2x_LOP868_238" & chrom == "chr01")) %>%
#   filter(!(sample == "2x_LOP868_259" & chrom == "chr02")) %>%
#   filter(!(sample == "2x_LOP868_292" & chrom == "chr02")) %>%
#   filter(!(sample == "2x_LOP868_452" & chrom == "chr02")) %>%
#   filter(!(sample == "2x_LOP868_460" & chrom == "chr02")) %>%
#   filter(!(sample == "2x_LOP868_1157" & chrom == "chr02")) %>%
#   filter(!(sample == "2x_LOP868_1190" & chrom == "chr02")) %>%
#   filter(!(sample == "2x_LOP868_373" & chrom == "chr03")) %>%
#   filter(!(sample == "2x_LOP868_028" & chrom == "chr04")) %>%
#   filter(!(sample == "2x_LOP868_433" & chrom == "chr04")) %>%
#   filter(!(sample == "2x_LOP868_293" & chrom == "chr05")) %>%
#   filter(!(sample == "2x_LOP868_363" & chrom == "chr05")) %>%
#   filter(!(sample == "2x_LOP868_172" & chrom == "chr06")) %>%
#   filter(!(sample == "2x_LOP868_065" & chrom == "chr07")) %>%
#   filter(!(sample == "2x_LOP868_272" & chrom == "chr07")) %>%
#   filter(!(sample == "2x_LOP868_108" & chrom == "chr08")) %>%
#   filter(!(sample == "2x_LOP868_488" & chrom == "chr08")) %>%
#   filter(!(sample == "2x_LOP868_397" & chrom == "chr10")) %>%
#   filter(!(sample == "2x_LOP868_373" & chrom == "chr02")) %>%
#   filter(!(sample == "2x_LOP868_128" & chrom == "chr12"))
```

```{r}
# a more elegant split-apply-combine approach
# see here for inspiration: http://www.stat.wisc.edu/~yandell/R_for_data_sciences/curate/tidyverse.html
dosage_geno_counts <- dosage_genos %>% 
  count(bin, cluster, dosage.gt) %>% 
  filter(n > 3) %>% 
  filter(n < 164)
head(dosage_geno_counts)
```

```{r}
dosage_geno_comps <- dosage_geno_counts %>% 
  setNames(paste0("target.", names(.))) %>% 
  crossing(dosage_geno_counts) %>%
  filter(target.bin != bin & target.dosage.gt != dosage.gt) %>%
  mutate(bincomp = ifelse(target.bin > bin, paste(target.bin, target.dosage.gt, bin, dosage.gt, sep = "_"), paste(bin, dosage.gt, target.bin, target.dosage.gt, sep = "_"))) %>% 
  distinct(bincomp, .keep_all = T) # here, .keep_all=T specifies to keep all variables
head(dosage_geno_comps) 
```

```{r}
dosage_geno_comps$inds.in.target <- rep(NA, nrow(dosage_geno_comps))
dosage_geno_comps$inds.out.target <- rep(NA, nrow(dosage_geno_comps))

# takes a goddamn year to run, how to speed this up?
ptm <- proc.time()
for (i in 1:nrow(dosage_geno_comps)) {
# for (i in 1:10) {
  source.bin <- dosage_geno_comps$bin[i]
  source.allele <- dosage_geno_comps$dosage.gt[i]
  target.bin <- dosage_geno_comps$target.bin[i]
  target.allele <- dosage_geno_comps$target.dosage.gt[i]
  inds <- filter(dosage_genos, bin == source.bin & dosage.gt == source.allele)$sample
  con <- filter(dosage_genos, sample %in% inds & bin == target.bin)
  dosage_geno_comps$inds.in.target[i] <- nrow(filter(con, dosage.gt == target.allele))
  dosage_geno_comps$inds.out.target[i] <- nrow(filter(con, dosage.gt != target.allele))
}
proc.time() - ptm
```

> Very slow. Within chromosome comparison takes about 4 minutes. This is going to scale exponentially with
the number of bins and alleles under comparison.

```{r}
dosage_geno_counts_split <- split(dosage_geno_counts, f=dosage_geno_counts$bin)
bin_allele_counts <- unlist(lapply(dosage_geno_counts_split, function(x) nrow(x)))
bin_allele_counts <- data.frame("target.bin" = names(bin_allele_counts),
           "target.allele.count" = bin_allele_counts)
```

```{r}
blah <- left_join(dosage_geno_comps, bin_allele_counts, by = "target.bin") %>% 
  mutate(target.expect = round((inds.in.target + inds.out.target) / target.allele.count))
```

```{r}
blah$fisher.pval <- apply(blah, 1, function(x) {
  tbl <- matrix(as.numeric(c(x[10], x[11], x[13], x[13])) , ncol = 2, byrow = T)
  fisher.test(tbl)$p.value
})
```

```{r}
# finally, add numeric bin back
blah$src.bin.num <- rep(NA, nrow(blah))
blah$target.bin.num <- rep(NA, nrow(blah))
dg.bins <- unique(dosage_genos$bin)
for (i in 1:nrow(blah)) {
  blah$src.bin.num[i] <- which(dg.bins == blah$bin[i])
  blah$target.bin.num[i] <- which(dg.bins == blah$target.bin[i])
}
```

```{r}
# for graphing purposes, may want to check src.bin.num and target.bin.num, and reverse source-target so that all tiles show up on the same side of the diagonal
# test this using a single chromosome
# source plot always greater of src.bin.num and target.bin.num
# target plot always lesser of src.bin.num and target.bin.num

# also add Benjamini-Hochberg FDR correction of pvalues
# rank P values
# Determine critical BH p-value (i/m)*Q where
# i is the rank of the p-value at hand
# m is the total number of tests
# Q is the chosen false discovery rate
# the largest P value that has P < (i/m)*Q is significant, and all other smaller P values are significant. All others are not.

fdr.rate <- 0.05
ntests <- nrow(blah)
blah <- blah %>%
  mutate(src.bin.num.plot = ifelse(src.bin.num > target.bin.num, src.bin.num, target.bin.num)) %>%  
  mutate(target.bin.num.plot = ifelse(src.bin.num > target.bin.num, target.bin.num, src.bin.num)) %>% 
  # mutate(fisher.pval.plot = ifelse(fisher.pval < 0.001, fisher.pval, NA))
  arrange(fisher.pval) %>% 
  rowid_to_column("fisher.pval.rank") %>% 
  mutate(fdr.bh.corrected = fdr.rate * fisher.pval.rank / ntests)
```

```{r}
# make glorious figure
# p <- ggplot(blah, aes(x = src.bin.num, y = target.bin.num, color = fisher.pval)) +
#   geom_tile()
# p
q <- ggplot(filter(blah, fdr.bh.corrected <= 0.005), aes(x = src.bin.num.plot, y = target.bin.num.plot)) +
  geom_tile()
q
# ggsave("2019_1116_fishermtx.pdf", width = 6, height = 6, units = "in", device = "pdf")
```
